auth/hash/hash.go

```diff
@@ -0,0 +1,203 @@
+}
+
+// StandardScrypt represents the standard scrypt hash algorithm.
+type StandardScrypt struct {
+	BlockSize        int
+	DerivedKeyLength int
+	MemoryCost       int
+	Parallelization  int
+}
+
+// Config returns the validated hash configuration.
+func (s *StandardScrypt) Config() (*internal.HashConfig, error) {
+	return &internal.HashConfig{
+		HashAlgorithm:    "STANDARD_SCRYPT",
+		DerivedKeyLength: int64(s.DerivedKeyLength),
+		BlockSize:        int64(s.BlockSize),
+		Parallelization:  int64(s.Parallelization),
+		MemoryCost:       int64(s.MemoryCost),
+		ForceSendFields:  []string{"BlockSize", "Parallelization", "MemoryCost", "DkLen"},
+	}, nil
+}
+
+// Scrypt represents the scrypt hash algorithm.
+//
+// This is the modified scrypt used by Firebase Auth (https://github.com/firebase/scrypt).
+type Scrypt struct {
+	Key           []byte
+	SaltSeparator []byte
+	Rounds        int
+	MemoryCost    int
+}
+
+// Config returns the validated hash configuration.
+func (s *Scrypt) Config() (*internal.HashConfig, error) {
+	if len(s.Key) == 0 {
+		return nil, errors.New("signer key not specified")
+	}
+	if s.Rounds < 1 || s.Rounds > 8 {
+		return nil, errors.New("rounds must be between 1 and 8")
+	}
+	if s.MemoryCost < 1 || s.MemoryCost > 14 {
+		return nil, errors.New("memory cost must be between 1 and 14")
+	}
+	return &internal.HashConfig{
+		HashAlgorithm: "SCRYPT",
+		SignerKey:     base64.RawURLEncoding.EncodeToString(s.Key),
+		SaltSeparator: base64.RawURLEncoding.EncodeToString(s.SaltSeparator),
+		Rounds:        int64(s.Rounds),
+		MemoryCost:    int64(s.MemoryCost),
+	}, nil
+}
+
+// HMACMD5 represents the HMAC SHA512 hash algorithm.
+type HMACMD5 struct {
+	Key []byte
+}
+
+// Config returns the validated hash configuration.
+func (h *HMACMD5) Config() (*internal.HashConfig, error) {
+	return hmacConfig("HMAC_MD5", h.Key)
+}
+
+// HMACSHA1 represents the HMAC SHA512 hash algorithm.
+type HMACSHA1 struct {
+	Key []byte
+}
+
+// Config returns the validated hash configuration.
+func (h *HMACSHA1) Config() (*internal.HashConfig, error) {
+	return hmacConfig("HMAC_SHA1", h.Key)
+}
+
+// HMACSHA256 represents the HMAC SHA512 hash algorithm.
+type HMACSHA256 struct {
+	Key []byte
+}
+
+// Config returns the validated hash configuration.
+func (h *HMACSHA256) Config() (*internal.HashConfig, error) {
+	return hmacConfig("HMAC_SHA256", h.Key)
+}
+
+// HMACSHA512 represents the HMAC SHA512 hash algorithm.
+type HMACSHA512 struct {
+	Key []byte
+}
+
+// Config returns the validated hash configuration.
+func (h *HMACSHA512) Config() (*internal.HashConfig, error) {
+	return hmacConfig("HMAC_SHA512", h.Key)
+}
+
+// MD5 represents the MD5 hash algorithm.
+type MD5 struct {
+	Rounds int
+}
+
+// Config returns the validated hash configuration.
+func (h *MD5) Config() (*internal.HashConfig, error) {
+	return basicConfig("MD5", h.Rounds)
+}
+
+// PBKDF2SHA256 represents the PBKDF2SHA256 hash algorithm.
```

Comment:
These comments feel unhelpful. Is there anything more reasonable we could put? Like maybe a link to more info about choosing a hash algorithm?

Reply:
Added links to the public documentation. It's still not perfect, but way better than just spelling out the type names. We can work with the auth team to get the documentation improved further.